### ISSUE TITLE
Concurrency issue

### ISSUE THREAD

**Issue by ionut-arm** [opendnssec/SoftHSMv2#573]
Hello!

We've been using SoftHSM2 for testing  - thank you very much for this library, it's been very useful to us - but recently we've started getting random failures that seem concurrency-related.

Our tests do various operations, but most of them rely on `C_GenerateKeyPair` or `C_CreateObject` to begin with. These are the two that ultimately fail (with `CKR_GENERAL_ERROR` or `CKR_TEMPLATE_INCONSISTENT`) - however which call to generate/create fails is pretty much random. The failed tests pass when run independently or even when rerunning the same call immediately after a failure. They only fail when run in a multi-threaded way. Within the service we create a new PKCS11 session for each operation and that session is only used in one thread. The library is initialized with `CKF_OS_LOCKING_OK`.

We generally use RSA key pairs, and from syslogs I could see that the failure usually comes from `P11KeyObj::init`, `P11PublicKeyObj::init` or `P11PrivateKeyObj::init`.

Digging deeper and adding some logging, I found that the error surfaces from [`ObjectFile::setAttribute`](https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/object_store/ObjectFile.cpp#L208) - `valid` ends up being false after the call to `store()`. The strange thing is that while logging at all possible return points of [`ObjectFile::store`](https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/object_store/ObjectFile.cpp#L661) I could never see `valid` turning false...

I'm not really sure what else to try here, or whether this is because of how we handle multi-threading or some bug somewhere else in the stack. 

Just as a note, we access SoftHSM2 through the [`pkcs11`](https://crates.io/crates/pkcs11) Rust crate.

**Comment by rijswijk** [comment]
Hmm, this is very tricky for us to figure out based on the information provided, would you be able to produce a minimally reproducing piece of example code? Also: which OS (distribution), crypto backend and crypto library version are you using?

**Comment by ionut-arm** [comment]
I've made a demo, but it's written in Rust - if you need to install the framework for it you can find the instructions [here](https://www.rust-lang.org/tools/install); then it's just a matter of doing a `cargo run` from the directory with `Cargo.toml`. You'll also need to change the constants at the top in `src/main.rs` to point to your SoftHSM2 library, your local slot number and your PIN.

The example spins up 6 threads which each will create and then immediately delete 1000 keys. Before each operation (creating/deleting one key) a new session is created. The parameters for the keys are hardcoded, apart from the key ID which is unique (generated at random).

After running for some amount of time (depending on the platform), an error pops up:
> thread '<unnamed>' panicked at 'Failed to generate key: Pkcs11(5)', src/main.rs:76:10

[error-demo.zip](https://github.com/opendnssec/SoftHSMv2/files/5194699/error-demo.zip)


> which OS 

I'm running WSL locally (`4.4.0-18362-Microsoft GNU/Linux`) - Ubuntu 18.04 to be more precise, but we've run the code on other platforms as well: our CI also uses Ubuntu 18.04 and we also ran it on Fedora release 32.

> crypto backend

I'm not sure about this, since it doesn't get configured at any point when installing the library - I assume it's OpenSSL. For our CI we only install `libssl-dev` in the container.

> library version 

In CI we use 2.5.0, locally I've tried 2.6.1 and the tip of `develop` as well.

All of them resulted in the same issue.

**Comment by rijswijk** [comment]
Thanks for the code that reproduces the issue, I've spent some time looking at it. So my current hypothesis is that there is a race condition between `C_GenerateKeyPair` and `C_FindObjects` in another thread. This causes the internal cache of the token to be trashed, because `C_FindObjects` causes a re-read of the data on disk, finds an empty object in the process of being created and marks it as invalid. This is probably because of the way locking is now implemented, the individual files are locked, but not the entire token directory. This may take quite a bit of time to fix, because it requires re-testing, I need to think about the impact of that first.

The good news is that this is highly specific to the use case you report where competing threads create keys while another thread is enumerating and deleting objects.

**Comment by ionut-arm** [comment]
Thanks, that does sound like a likely culprit! I did some `gdb` sessions into the SoftHSM2 code but didn't know it well enough to figure out the exact cause. 
I did, however, notice some possible locking issues, but couldn't tell whether they're as intended or oversights - for example [here](https://github.com/opendnssec/SoftHSMv2/blob/develop/src/lib/object_store/ObjectFile.cpp#L327) `objectFile` is locked on L319 but isn't unlocked before this return (whereas it is for all subsequent returns).

**Comment by rijswijk** [comment]
Hmm, yes, there should probably be a call to `unlock` there, but this is not the cause of the issue you are observing, I'm now 99.9% certain it is a race between object creation and enumerating objects in separate threads.

**Comment by rijswijk** [comment]
OK, so it looks like there may be a trivial fix: right where you pointed out a call to `unlock` is missing, do two things:

Commented out `valid = false;`
Add the call to `unlock`

Can you try this and confirm it works for you? It seems to work with your example code. If this fixes your issue, my next step is to re-run regression tests.

**Comment by ionut-arm** [comment]
I've tried and it worked for me locally, but @hug-dev also tried locally and for him there were some issues - not exactly like before, and there might have been some other unrelated problems involved there.

I'd be happy for the regression tests to be run, I think there's a good chance for this to solve our problem.

Just for context, we use SoftHSM2 for CI testing and since we discovered this issue we essentially locked our PKCS11 interface to be single-threaded so this kind of error doesn't occur, as we didn't know whether this is a fault on our side or not (plus, having the CI always failing is not a good sign). It's by no means critical/urgent on our side, though. There might be other users in a similar position out there, maybe (?)

Thank you again for looking into this!
